base/broadcast.jl

```diff
@@ -253,17 +250,41 @@ function broadcast_t(f, ::Type{Any}, As...)
     B[I] = val
     return _broadcast!(f, B, keeps, Idefaults, As, Val{nargs}, iter, st, 1)
 end
-
-@inline broadcast_t(f, T, As...) = broadcast!(f, similar(Array{T}, broadcast_indices(As...)), As...)
+@inline function _broadcast_t(f, T, shape, iter, As...)
+    B = similar(Array{T}, shape)
+    nargs = length(As)
+    keeps, Idefaults = map_newindexer(shape, As)
+    _broadcast!(f, B, keeps, Idefaults, As, Val{nargs}, iter)
+    return B
+end
+@inline broadcast_t(f, As...) = broadcast!(f, similar(Array{promote_eltype_op(f, As...)}, broadcast_indices(As...)), As...)
+
+ftype(f, A) = typeof(a->f(a))
+ftype(f, A...) = typeof(a->f(a...))
+ftype(T::DataType, A) = Type{T}
+ftype(T::DataType, A...) = Type{T}
+ziptype(A) = Tuple{eltype(A)}
+ziptype(A, B) = Zip2{Tuple{eltype(A)},Tuple{eltype(B)}}
+@inline ziptype(A, B, C, D...) = Zip{Tuple{eltype(A)},ziptype(B, C, D...)}
+
+@inline function _broadcast(f, As...)
+    S = _default_eltype(Generator{ziptype(As...),ftype(f, As...)})
+    shape = broadcast_indices(As...)
+    iter = CartesianRange(shape)
+    if isempty(iter)
+        return similar(Array{S}, shape)
+    end
+    T = isleaftype(S) ? S : Any
+    return _broadcast_t(f, T, shape, iter, As...)
+end
 
 function broadcast_c(f, ::Type{Tuple}, As...)
     shape = broadcast_indices(As...)
-    check_broadcast_indices(shape, As...)
```

Comment:
`broadcast_indices` already fails when the sizes are not compatible. The place where we really need it is for `broadcast!` where we don't know the size of the supplied container.